src/game.ts

```diff
@@ -353,59 +366,31 @@ const buildTx = async (
   datum: string,
   collateralUtxo: UTxO,
 ): Promise<[UTxO, TxSigned]> => {
-  // HACK: hydra returns a decoded/json format of the datum, so we coerce it back here
-  let lucidUtxo = {
-    txHash: inputUtxo.txHash,
-    outputIndex: inputUtxo.outputIndex,
-    address: inputUtxo.address,
-    assets: inputUtxo.assets,
-    datum:
-      !!inputUtxo.datum && typeof inputUtxo.datum !== "string"
-        ? Data.to(hydraDatumToPlutus(inputUtxo.datum))
-        : inputUtxo.datum,
-    datumHash: inputUtxo.datumHash,
-    scriptRef: inputUtxo.scriptRef,
-  };
-
-  const tx = lucid.newTx();
-  tx.collectFrom([lucidUtxo], redeemer);
-
-  tx.payToContract(scriptAddress, { inline: datum }, { lovelace: BigInt(0) });
-  tx.readFrom([
-    {
-      txHash: scriptRef!.split("#")[0],
-      outputIndex: Number(scriptRef!.split("#")[1]),
-      address: scriptAddress,
-      assets: { lovelace: BigInt(0) },
-      scriptRef: {
-        type: "PlutusV2",
-        script: CBOR,
-      },
-    },
-  ]);
-  tx.addSigner(address);
-  let complete;
-  try {
-    complete = await tx.complete();
-  } catch (e) {
-    throw e;
-  }
-  const collateral = buildCollateralInput(
-    collateralUtxo.txHash,
-    collateralUtxo.outputIndex,
+  // Hand-roll transaction creation for more performance
+  // NOTE: Redeemer is always using max ex units
+  const redeemerBlock = `81840000${redeemer}821a00d59f801b00000002540be400`;
+  const datumBlock = ``; // No datums in the witness set, so this is empty string
+  const scriptData = `${redeemerBlock}${datumBlock}${COST_MODELS}`;
+  const scriptDataHash = bytesToHex(
+    blake2b(hexToBytes(scriptData), { dkLen: 256 / 8 }),
   );
-  const txBody = complete.txComplete.body();
-  const witnessSet = complete.txComplete.witness_set();
-  const auxData = complete.txComplete.auxiliary_data();
-
-  txBody.set_collateral(collateral);
-  const collateralTx = C.Transaction.new(txBody, witnessSet, auxData);
-  txBody.free();
-  witnessSet.free();
-  auxData?.free();
-  collateral.free();
-  complete.txComplete = collateralTx;
-  const signedTx = await complete.sign().complete();
+  const datumLength = datum.length / 2;
+  const txBodyByHand =
+    `a7` + // Prefix
+    `0081825820${inputUtxo.txHash}0${inputUtxo.outputIndex}` + // One input
+    `0181a300581d70${scriptHash}018200a0028201d81858${datumLength.toString(16)}${datum}` + // Output to script hash with datum
+    `0200` + // No fee
+    `0b5820${scriptDataHash}` + // Script data hash, spooky
+    `0d81825820${collateralUtxo.txHash}0${collateralUtxo.outputIndex}` + // Collatteral Input
+    `0e81581c${pkh}` + // Required Signers
+    `1281825820${scriptRef!.split("#")[0]}0${scriptRef!.split("#")[1]}`; // Reference inputs
+  const witnessSetByHand = `a105${redeemerBlock}`; // a single redeemer in witness set
+  const txByHand = `84${txBodyByHand}${witnessSetByHand}f5f6`;
+
+  // Still use lucid for signing with configured key
+  const tx = lucid.fromTx(txByHand);
```

Comment:
in theory we can even use the @noble library to sign this directly; probably not worth it if we're down to 2-3ms, but i'm surprised the overhead of parsing the tx isn't higher.